### PR TITLE
ENH: add vector geometries

### PR DESCRIPTION
**Pull request by kohr-h**

### CODE REVIEW

File: examples/tomo/ray_trafo_vec_geom_3d.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+"""Example using the ray transform a custom vector geometry.
```
Comment (kohr-h) [inline examples/tomo/ray_trafo_vec_geom_3d.py]: with

File: odl/tomo/backends/astra_setup.py [hunk 2]
```diff
@@ -424,6 +482,22 @@ def astra_projection_geometry(geometry):
...
+        if not astra_supports('parallel2d_vec_geometry'):
+            raise NotImplementedError(
+                "'parallel_vec' geometry not supported by ASTRA "
+                'v{}'.format(ASTRA_VERSION))
+        proj_geom = astra.create_proj_geom('parallel_vec', det_count, vec)
+
+    elif isinstance(geometry, ConeVecGeometry) and geometry.ndim == 2:
+        det_count = geometry.detector.size
+        vec = geometry.vectors
+        # TODO: flip axes?
```
Comment (kohr-h) [inline odl/tomo/backends/astra_setup.py]: no

File: odl/tomo/backends/astra_setup.py [hunk 3]
```diff
@@ -441,6 +515,23 @@ def astra_projection_geometry(geometry):
         vec = astra_conebeam_3d_geom_to_vec(geometry)
         proj_geom = astra.create_proj_geom('cone_vec', det_row_count,
                                            det_col_count, vec)
+
+    elif isinstance(geometry, ParallelVecGeometry) and geometry.ndim == 3:
+        det_row_count = geometry.det_partition.shape[1]
+        det_col_count = geometry.det_partition.shape[0]
+        vec = geometry.vectors
+        # TODO: flip axes?
```
Comment (kohr-h) [inline odl/tomo/backends/astra_setup.py]: no

File: odl/tomo/backends/astra_setup.py [hunk 4]
```diff
@@ -441,6 +515,23 @@ def astra_projection_geometry(geometry):
...
+        proj_geom = astra.create_proj_geom('parallel3d_vec', det_row_count,
+                                           det_col_count, vec)
+
+    elif isinstance(geometry, ConeVecGeometry) and geometry.ndim == 3:
+        det_row_count = geometry.det_partition.shape[1]
+        det_col_count = geometry.det_partition.shape[0]
+        vec = geometry.vectors
+        # TODO: flip axes?
```
Comment (kohr-h) [inline odl/tomo/backends/astra_setup.py]: no

File: odl/tomo/geometry/geometry.py [hunk 5]
```diff
@@ -366,6 +369,420 @@ def rotation_matrix(self, angle):
...
+        axis : `numpy.ndarray`, shape ``(2,)``
+            The detector axis at ``index``.
+
+        Examples
+        --------
+        Initialize a 2D parallel beam geometry with detector axes
+        ``(1, 0)`` and ``(0, 1)`` (columns 4 and 5 in the vectors,
+        counting from 0):
+
+        >>> # d = refpoint
```
Comment (kohr-h) [inline odl/tomo/geometry/geometry.py:715]: fix

File: odl/tomo/geometry/geometry.py [hunk 6]
```diff
@@ -366,6 +369,420 @@ def rotation_matrix(self, angle):
...
+            The detector axes at ``index``.
+
+        Examples
+        --------
+        Initialize a 3D parallel beam geometry with detector axis ``u``
+        vectors ``(1, 0, 0)`` and ``(0, 1, 0)`` (columns 6 through 8 in the
+        vectors, counting from 0), and axis ``v`` equal to ``(0, 0, 1)``
+        in both cases (columns 9 through 11):
+
+        >>> # d = refpoint
```
Comment (kohr-h) [inline odl/tomo/geometry/geometry.py:772]: fix

File: odl/tomo/operators/ray_trafo.py [hunk 7]
```diff
@@ -204,18 +205,26 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
             dtype = reco_space.dtype
             proj_uspace = FunctionSpace(geometry.params, out_dtype=dtype)
 
+            # TODO: use weighting that differentiates between angles and
+            # detector. We need an array of weights here, a different value
+            # for each angle, but constant in the detector axes.
+            # The required partition property is available since
+            # commit a551190d, but weighting is not adapted yet.
+            # See also issues #286 and #1001
+            if isinstance(self.geometry, ParallelVecGeometry):
```
Comment (kohr-h) [inline odl/tomo/operators/ray_trafo.py]: `VecGeometry`

**Comment by kohr-h** [comment]
It will be hard to get the adjoint entirely correct with this setting unless we have 1D array weighting.

**Comment by adler-j** [comment]
Are you talking to yourself here? Or should I help?

**Comment by kohr-h** [comment]
Largely talking to myself :-)

This will become a PR into main repo once the geometry PR is in.

**Pull request closed without merging**